### PR TITLE
Adds Video Surfer to autogen_ext

### PR DESCRIPTION
**Pull request by gagb**
This pull request introduces a new `VideoSurferAgent` and its associated tools to the `autogen-ext` package, while also making some adjustments to existing agents. The main changes include the addition of the `VideoSurferAgent`, updates to the `MultimodalWebSurfer` import paths, and the removal of the `MultimodalWebSurfer` from the `__init__.py` of `agents`.

### New `VideoSurferAgent` and tools:

* [`python/packages/autogen-ext/pyproject.toml`](diffhunk://#diff-095119d4420ff09059557bd25681211d1772c2be0fbe0ff2d551a3726eff1b4bR34-R40): Added dependencies for the new `VideoSurferAgent`.
* [`python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/__init__.py`](diffhunk://#diff-756d809674f8f3fad2b2190fb80605239347c72a408aae14ba1680b19f857b43R1-R7): Introduced the `VideoSurferAgent` and its tools submodule.
* [`python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_video_surfer.py`](diffhunk://#diff-f9ce9a97dff9d0dee184ba3a896041374888f1e70e0cc15c8bf4ca8e9c889169R1-R150): Implemented the `VideoSurferAgent` class, which can answer questions about local video files using various tools.
* [`python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/tools.py`](diffhunk://#diff-f37f774d6fde0d233c4ce0db9fb0a80df2397617742b9793173704651fec43d9R1-R162): Added tools for the `VideoSurferAgent` to extract audio, transcribe audio, get video length, save screenshots, and transcribe video screenshots using OpenAI API.

### Adjustments to existing agents:

* [`python/packages/autogen-ext/src/autogen_ext/agents/__init__.py`](diffhunk://#diff-642304905f1bd9913dc6ebe3ec1fd47dcbaee722d0e6f96a108acf439f9138abL2-R3): Removed the `MultimodalWebSurfer` import and adjusted the `__all__` list accordingly.
* [`python/packages/autogen-ext/src/autogen_ext/agents/web_surfer/__init__.py`](diffhunk://#diff-229e28d56688ecefb971118579e0c181dd9f497b8c22b1efdfecd7b5260c7d42R1-R3): Reintroduced the `MultimodalWebSurfer` in its own `__init__.py`.
* [`python/packages/autogen-studio/autogenstudio/database/component_factory.py`](diffhunk://#diff-b06aa9da31daf6c82cf8d8623952c27c8873b6127a04dbd285c611dc60a0daecL13-R13): Updated the import path for `MultimodalWebSurfer`.

### CODE REVIEW

File: python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_video_surfer.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+from autogen_agentchat.agents import AssistantAgent
+
+from ._action_space import extract_audio, get_video_length, transcribe_audio_with_timestamps, get_screenshot_at, save_screenshot, openai_transcribe_video_screenshot, download_youtube_video
+
+
+class VideoSurferAgent(AssistantAgent):
```
Comment (ekzhu) [inline python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_video_surfer.py:6]: Use class doc string for the documentation instead of the README.md. Much easier to discover.

File: python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_video_surfer.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    def __init__(self, *args, **kwargs):
```
Comment (ekzhu) [inline python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_video_surfer.py:7]: Explicitly set the argument list here for better readability. Also, set `system_message` default at the argument level.

File: python/packages/autogen-ext/src/autogen_ext/agents/__init__.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
 from ._openai_assistant_agent import OpenAIAssistantAgent
-from .web_surfer._multimodal_web_surfer import MultimodalWebSurfer
 
-__all__ = ["MultimodalWebSurfer", "OpenAIAssistantAgent"]
```
Comment (ekzhu) [inline python/packages/autogen-ext/src/autogen_ext/agents/__init__.py:4]: Why not keep all agents at the top-level?

Comment (gagb) [reply python/packages/autogen-ext/src/autogen_ext/agents/__init__.py:4]: The issue was that different agents have different dependencies, and this was causing for all to be imported?

Comment (ekzhu) [reply python/packages/autogen-ext/src/autogen_ext/agents/__init__.py:4]: Yes. You are right!

File: python/packages/autogen-ext/pyproject.toml [hunk 4]
```diff
@@ -31,6 +31,13 @@ web-surfer = [
     "playwright>=1.48.0",
     "pillow>=11.0.0",
 ]
+video-surfer = [
+    "autogen-agentchat==0.4.0.dev7",
```
Comment (ekzhu) [inline python/packages/autogen-ext/pyproject.toml:35]:
```suggestion
    "autogen-agentchat==0.4.0.dev8",
```

File: python/packages/autogen-core/docs/src/user-guide/agentchat-user-guide/assignments/video-surfer.md
Comment (ekzhu) [inline python/packages/autogen-core/docs/src/user-guide/agentchat-user-guide/assignments/video-surfer.md:1]: Let's make a separate PR for the assignment.

File: python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_video_surfer.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+from ._action_space import extract_audio, get_video_length, transcribe_audio_with_timestamps, get_screenshot_at, save_screenshot, openai_transcribe_video_screenshot
+
+
+class VideoSurferAgent(AssistantAgent):
+    """
+    VideoSurferAgent is a specialized agent designed to answer questions about a local video file.
+
+    This agent utilizes various tools to extract information from the video, such as its length, screenshots at specific timestamps, and audio transcriptions. It processes these elements to provide detailed answers to user queries.
+
```
Comment (ekzhu) [inline python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_video_surfer.py:15]: Need an argument list here to explain each argument. Currently for `tools` the behavior is overriding if supplied, and by default all the video tools from the action space are included. 

We should also list the tools in this doc so devs understand what tools are available. Use Sphinx RST reference syntax to reference the tools.

File: python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/__init__.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+from ._video_surfer import VideoSurferAgent
+
+__all__ = ["VideoSurferAgent"]
```
Comment (ekzhu) [inline python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/__init__.py:3]: Shall we also make available the tools. These tools can also be used by other agents. Exporting them also makes it possible to reference them from VideoSurferAgent's doc string.

File: python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_action_space.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
+import base64 
+import ffmpeg
+import cv2
+import whisper
+import openai
+
+
+def extract_audio(video_path: str, audio_output_path: str) -> str:
```
Comment (ekzhu) [inline python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_action_space.py:8]: To be included in a separate PR, but I believe it is essential to create unit tests for these tools. Consider data generation approach to generate tiny videos that can be used for tests.

File: python/packages/autogen-ext/pyproject.toml [hunk 8]
```diff
@@ -42,7 +49,7 @@ dev-dependencies = []
 [tool.ruff]
 extend = "../../pyproject.toml"
 include = ["src/**", "tests/*.py"]
-exclude = ["src/autogen_ext/agents/web_surfer/*.js"]
+exclude = ["src/autogen_ext/agents/web_surfer/*.js", "src/autogen_ext/agents/video_surfer/_action_space.py"]
```
Comment (ekzhu) [inline python/packages/autogen-ext/pyproject.toml:52]: 🤔

Comment (gagb) [reply python/packages/autogen-ext/pyproject.toml:52]: How do I stop it from complaining about type errors in a file that contains mostly external libs?

File: python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/__init__.py (hunk 6 above)
Comment (gagb) [reply python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/__init__.py:3]: Good point.

File: python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_action_space.py (hunk 7 above)
Comment (gagb) [reply python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_action_space.py:8]: okay!

File: python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_video_surfer.py [hunk 9]
```diff
@@ -13,6 +13,15 @@ class VideoSurferAgent(AssistantAgent):
 
     This agent utilizes various tools to extract information from the video, such as its length, screenshots at specific timestamps, and audio transcriptions. It processes these elements to provide detailed answers to user queries.
 
+    Available tools:
+    
+    - :func:`~autogen_ext.agents.video_surfer._action_space.extract_audio`
```
Comment (ekzhu) [inline python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/_video_surfer.py:25]: We can export the tools as public module? `_action_space` implies private module.

**Comment by gagb** [comment]
@victordibia, FYI there is a small change in studio's `component_factory.py` in this PR, in particular the import.

File: python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/tools.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    if not screenshots:
+        return "Failed to capture screenshot."
+
+    _, frame = screenshots[0]
+    # Convert the frame to bytes and then to base64 encoding
+    _, buffer = cv2.imencode(".jpg", frame)
+    frame_bytes = buffer.tobytes()
+    frame_base64 = base64.b64encode(frame_bytes).decode("utf-8")
+
+    client = openai.Client()
```
Comment (husseinmozannar) [inline python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/tools.py:103]: why are we setting the default of this to gpt-4o-mini instead of using the model client of the video surfer?

**Review by ekzhu — APPROVED** [review]
Let's create an issue to track the unit tests for tools.

File: python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/tools.py (hunk 10 above)
Comment (gagb) [reply python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/tools.py:103]: fixing it. good catch.

Comment (gagb) [reply python/packages/autogen-ext/src/autogen_ext/agents/video_surfer/tools.py:103]: submitted a new pr.
